src/main/java/com/salesforce/dynamodbv2/mt/util/CachingAmazonDynamoDbStreams.java

```diff
@@ -724,14 +531,83 @@ public GetRecordsResult getRecords(GetRecordsRequest request) {
             LOG.debug("getRecords request={}", request);
         }
 
+        final String iteratorString = request.getShardIterator();
+        checkArgument(iteratorString != null);
+        final int limit = Optional.ofNullable(request.getLimit()).orElse(GET_RECORDS_LIMIT);
+        checkArgument(limit > 0 && limit <= GET_RECORDS_LIMIT);
+
         // parse iterator
         final CachingShardIterator iterator = CachingShardIterator.fromExternalString(request.getShardIterator());
 
-        // fetch records using cache
-        final GetRecordsResult loadedResult = getRecords(iterator);
+        // Lock the shard to avoid (1) exceeding streams reader limit (See
+        // https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/Limits.html#limits-dynamodb-streams)
+        // and (2) concurrent readers from fetching overlapping segments unnecessarily
+        final ShardId shardId = iterator.getShardUid();
+        final Lock lock = shardLocks.get(shardId);
+        try {
+            if (!lock.tryLock(5000, TimeUnit.SECONDS)) {
+                if (LOG.isWarnEnabled()) {
+                    LOG.warn("getRecords failed to acquire lock for shard ", shardId);
+                }
+                throw new LimitExceededException("Failed to acquire shard lock within max time.");
+            }
+        } catch (InterruptedException e) {
+            if (LOG.isWarnEnabled()) {
+                LOG.warn("getRecords interrupted trying to acquire lock for shard ", shardId);
+            }
+            Thread.currentThread().interrupt();
+        }
+
+        final List<Record> records = new ArrayList<>(limit);
+        final CachingShardIterator nextIterator;
+        try {
+            // fetch records from cache
+            iterator.resolveLocation()
+                .map(location -> recordCache.getRecords(location, limit))
+                .ifPresent(records::addAll);
+
+            // check if we got enough
+            if (records.size() < limit) {
+                // not enough cached records: read segment from underlying shard
```

Comment:
the number of delta API requests that step into this block needs to be bubbled up to to ZOS so we can understand the ratio of queries that are served off this cache exclusively, to the ratio of queries that hit the underlying stream, to be reloaded, and the cost to updating the cache. 

Given this cache is homegrown, we can't rely on the Guava cache monitoring we have to get things like hitCount, loadTime, evictionCount, etc.. so we have to instrument those pieces ourselves. Given how critical this cache is to a performant delta API, we need those metrics

Reply:
Agree. There are also some more lower-level stats at the caching layer we may want to emit, like how many cache segments we iterated over to collect the result.

This is one of the functional changes compared to the version currently in master: in master the caching adapter returns immediately if it finds any number of records in the cache (see https://github.com/salesforce/mt-dynamo/blob/master/src/main/java/com/salesforce/dynamodbv2/mt/util/CachingAmazonDynamoDbStreams.java#L758). And the multitenant adapter continues to call `getRecords` if it got any back (https://github.com/salesforce/mt-dynamo/blob/master/src/main/java/com/salesforce/dynamodbv2/mt/mappers/sharedtable/impl/MtAmazonDynamoDbStreamsBySharedTable.java#L106).

With this change, the caching layer tries to return as many records as possible, by iterating over adjacent cache segments (https://github.com/salesforce/mt-dynamo/pull/424/files#diff-144e0b563c0b1615fa0417b18acc0778R194) and then makes a single attempt to fetch more records from the underlying stream if it didn't get enough. The multitenant adapter only continues to call `getRecords` if it retrieved the maximum number of records (https://github.com/salesforce/mt-dynamo/pull/424/files#diff-a48916ee1793548f6147598a5d2fcf1eR106).

DynamoDB docs actually state "GetRecords can retrieve a maximum of 1 MB of data or 1000 stream records, whichever comes first." (https://docs.aws.amazon.com/amazondynamodb/latest/APIReference/API_streams_GetRecords.html), so we may want/need to consider checking for 1MB or 1k.

Reply:
👍